### PR TITLE
feat(command): Implement Command GET /device/name/{name} V2 API

### LINKED ISSUE #3035: [V2 Command] Implement Command GET /device/name/{name} V2 API
# 🚀 Feature Request

### Relevant Package

 This feature request is for core command

### Description

Per https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-command/2.x#/default/get_device_name__name_, implement Command GET /device/name/{name} V2 API

### PR DESCRIPTION
**Pull request by judehung**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## Issue Number:
fix #3035 

## What is the new behavior?
Per https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-command/2.x#/default/get_device_name__name_, implement Command GET /device/name/{name} V2 API

Signed-off-by: Jude Hung <jude@iotechsys.com>

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by judehung** [comment]
rebased

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/core/command/v2/mockclients/interfaces/mocks/MockDeviceClient.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
+// Code generated by mockery v2.5.1. DO NOT EDIT.
+
+package mocks
```
Comment (lenny-intel) [inline internal/core/command/v2/mockclients/interfaces/mocks/MockDeviceClient.go:3]: These mocks should be in core-contracts so they can be used through out edgex.

File: internal/core/command/v2/mockclients/interfaces/device.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+//
+// Copyright (C) 2021 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package interfaces
+
+import "github.com/edgexfoundry/go-mod-core-contracts/v2/v2/clients/interfaces"
+
+type MockDeviceClient interface {
```
Comment (lenny-intel) [inline internal/core/command/v2/mockclients/interfaces/device.go:10]: Why are these needed. Mockery should create all that is required.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/core/command/v2/application/command.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	commandContainer "github.com/edgexfoundry/edgex-go/internal/core/command/container"
+	v2CommandContainer "github.com/edgexfoundry/edgex-go/internal/core/command/v2/bootstrap/container"
+	"github.com/edgexfoundry/go-mod-bootstrap/v2/di"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/errors"
+	V2Routes "github.com/edgexfoundry/go-mod-core-contracts/v2/v2"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/v2/dtos"
+)
+
+// CommandsByDeviceName query coreCommands with device name
+func CommandsByDeviceName(name string, dic *di.Container) (commands []dtos.CoreCommand, err errors.EdgeX) {
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/command/v2/application/command.go:21]: @lenny-intel I totally miss this part.  I just reviewed the source code and noticed core-command connects to database directly.
In the very original design / implementation (Java version), core-command doesn't access database and only communicate with core-metadata.  I thought it is always like this.
https://github.com/edgexfoundry/core-command/blob/master/src/main/java/org/edgexfoundry/controller/impl/CommandControllerImpl.java#L112-L115

@jpwhitemn please help add this topic to the WG meeting.  Thanks.

[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/core/command/v2/application/command.go:21]: @lenny-intel please see our official document, and there sequence diagrams to show the design
https://docs.edgexfoundry.org/1.3/microservices/core/command/Ch-Command/

File: internal/core/command/v2/bootstrap/container/metadata.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package container
+
+import (
+	"github.com/edgexfoundry/go-mod-bootstrap/v2/di"
+	v2Clients "github.com/edgexfoundry/go-mod-core-contracts/v2/v2/clients/interfaces"
+)
+
+// MetadataDeviceClientName contains the name of the Metadata device client instance in the DIC.
```
[the remark this answers is not in the document]
Comment (judehung) [reply internal/core/command/v2/bootstrap/container/metadata.go:13]: Agree.  I've submitted a PR at https://github.com/edgexfoundry/go-mod-bootstrap/pull/158 to create these helper functions .

File: internal/core/command/v2/mockclients/interfaces/mocks/MockDeviceClient.go (hunk 1 above)
Comment (judehung) [reply internal/core/command/v2/mockclients/interfaces/mocks/MockDeviceClient.go:3]: Agree.  I've submitted a PR at https://github.com/edgexfoundry/go-mod-core-contracts/pull/457 to create mocks at go-mod-core-contracts.

File: internal/core/command/v2/mockclients/interfaces/device.go (hunk 2 above)
Comment (judehung) [reply internal/core/command/v2/mockclients/interfaces/device.go:10]: The V2 client library interfaces are defined under go-mod-core-contracts, and Mockery doesn't directly support the feature to generate mocks from an external go modules.  My trick is to create corresponding Mock interfaces with inline fields directly referring to the V2 client library interface of go-mod-core-contracts, so that Mockery could generate the mocks within edgex-go.  However, as mocks have been moved to go-mod-core-contracts, I'll remove these mocking interfaces.

**Comment by judehung** [comment]
rebased and adjusted codes per Lenny's review comments.

**Review by lenny-intel — APPROVED** [review]
LGTM
